File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Ruby/lib/specific_bad_consequence.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'bad_consequence.rb'

class SpecificBadConsequence < BadConsequence
  
  attr_reader :specificVisibleTreasures, :specificHiddenTreasures
  
  def initialize(text,levels,specificVisibleTreasures,specificHiddenTreasures)
    super(text,levels)
    @specificVisibleTreasures = specificVisibleTreasures
    @specificHiddenTreasures = specificHiddenTreasures
  end
  
  def empty?
    (@specificVisibleTreasures.empty?) && (@specificHiddenTreasures.empty?)
  end
  
  def substractVisibleTreasure(t)
    if(@specificVisibleTreasures.include?(t.type)) then
      @specificVisibleTreasures.delete(t.type)
    end
  end
  
  def substractHiddenTreasure(t)
    if(@specificHiddenTreasures.include?(t.type)) then
      @specificHiddenTreasures.delete(t.type)
    end
  end
  
  def adjustToFitTreasureLists(v,h)
    v2 = Array.new  
    h2 = Array.new
    #ONEHAND ocultos
    nOneHand = 0
    for type in @specificHiddenTreasures do
      if(type == TreasureKind::ONEHAND)
        nOneHand = nOneHand + 1
      end
    end
    for treasure in h do
      if(@specificHiddenTreasures.include?(treasure.type) && treasure.type != TreasureKind::ONEHAND) then
        h2 << treasure.type
      elsif(@specificHiddenTreasures.include?(treasure.type) && (treasure.type == TreasureKinde::ONEHAND) && (nOneHand > 0)) then
        h2 << treasure.type  
        nOneHand = nOneHand - 1
      end
    end
    #ONEHAND visibles
    nOneHand = 0
    for type in @specificVisibleTreasures do
      if(type == TreasureKind::ONEHAND)
        nOneHand = nOneHand + 1
      end
    end
    for treasure in v do
      if(@specificVisibleTreasures.include?(treasure.type) && treasure.type != TreasureKind::ONEHAND) then
        v2 << treasure.type
      elsif(@specificVisibleTreasures.include?(treasure.type) && (treasure.type == TreasureKind::ONEHAND) && (nOneHand > 0)) then
        v2 << treasure.type  
        nOneHand = nOneHand - 1
      end
    end
    SpecificBadConsequence.new(@text,@levels,v2,h2)
  end
  
  def to_s
    super() + ", specificVisibleTreasures=#{@specificVisibleTreasures}, specificHiddenTreasures=#{@specificHiddenTreasures}"
  end
end

File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Java/src/NapakalakiGame/SpecificBadConsequence.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.ArrayList;

/**
 *
 * @author juanka1995
 */
public class SpecificBadConsequence extends BadConsequence {

    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    public SpecificBadConsequence(String text, int levels, 
            ArrayList<TreasureKind> tVisible, ArrayList<TreasureKind> tHidden) {
        super(text,levels);
        specificVisibleTreasures = tVisible;
        specificHiddenTreasures = tHidden;
    }
    
    @Override
    public boolean isEmpty(){
        return (specificHiddenTreasures.isEmpty() && specificVisibleTreasures.isEmpty());
    }
    
    public ArrayList<TreasureKind> getSpecificHiddenTreasures() {
        return specificHiddenTreasures;
    }

    public ArrayList<TreasureKind> getSpecificVisibleTreasures() {
        return specificVisibleTreasures;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t){
        if(specificVisibleTreasures.contains(t.getType()))
            specificVisibleTreasures.remove(t.getType());
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t){
        if(specificHiddenTreasures.contains(t.getType()))
            specificHiddenTreasures.remove(t.getType());
    }
    
    @Override
    public SpecificBadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v, ArrayList<Treasure> h){
        SpecificBadConsequence nuevo;
        int nOneHand = 0;
        if (!specificHiddenTreasures.isEmpty() || !specificVisibleTreasures.isEmpty()){
            ArrayList<TreasureKind> v2 = new ArrayList();
            ArrayList<TreasureKind> h2 = new ArrayList();
            // Calculo el numero de ONEHAND ocultos
            for ( TreasureKind type:this.specificHiddenTreasures )
                if(type == TreasureKind.ONEHAND)
                    nOneHand++;
            for ( Treasure treasure:h ) {
                if(specificHiddenTreasures.contains(treasure.getType()) && treasure.getType() != TreasureKind.ONEHAND){
                    h2.add(treasure.getType());
                }
                else if (specificHiddenTreasures.contains(treasure.getType()) && treasure.getType() == TreasureKind.ONEHAND && nOneHand > 0){
                    h2.add(treasure.getType());
                    nOneHand--;
                }
            }
            nOneHand = 0;
            // Calculo el numero de ONEHAND visibles
            for ( TreasureKind type:this.specificVisibleTreasures )
                if(type == TreasureKind.ONEHAND)
                    nOneHand++;
            for ( Treasure treasure:v ) {
                if(specificVisibleTreasures.contains(treasure.getType()) && treasure.getType() != TreasureKind.ONEHAND){
                    v2.add(treasure.getType());
                }
                else if (specificVisibleTreasures.contains(treasure.getType()) && treasure.getType() == TreasureKind.ONEHAND && nOneHand > 0){
                    v2.add(treasure.getType());
                    nOneHand--;
                }
            }
            nuevo = new SpecificBadConsequence(super.getText(), super.getLevels(), v2, h2);
        }
        else{
            return this;
        }
        return nuevo;
    }
    
    @Override
    public String toString() {
        return super.toString() + "\nspecificHiddenTreasures: " + specificHiddenTreasures + "\nspecificVisibleTreasures: " + specificVisibleTreasures;
    }
}
